agent/engine/task_manager.go

```diff
@@ -490,61 +527,89 @@ func (mtask *managedTask) handleEventError(containerChange dockerContainerChange
-		// clearly can't just continue trying to transition it to stopped
-		// again and again. In this case, assume it's stopped (or close
-		// enough) and get on with it
-		// This can happen in cases where the container we tried to stop
-		// was already stopped or did not exist at all.
-		seelog.Warnf("Managed task [%s]: 'docker stop' returned %s: %s",
-			mtask.Arn, event.Error.ErrorName(), event.Error.Error())
-		container.SetKnownStatus(api.ContainerStopped)
-		container.SetDesiredStatus(api.ContainerStopped)
-	} else if event.Status == api.ContainerPulled {
-		// Another special case; a failure to pull might not be fatal if e.g. the image already exists.
+	switch event.Status {
+	// event.Status is the desired container transition from container's known status
+	// (* -> event.Status)
+	case api.ContainerPulled:
+		// Container's desired transition was to 'PULLED'. A failure to pull might
+		// not be fatal if e.g. the image already exists.
 		seelog.Errorf("Managed task [%s]: Error while pulling container %s, will try to run anyway: %v",
 			mtask.Arn, container.Name, event.Error)
-	} else {
-		seelog.Warnf("Managed task [%s]: Error with docker; stopping container %s: %v",
+		// proceed anyway
+		return true
+	case api.ContainerStopped:
+		// Container's desired transition was to 'STOPPED'
+		return mtask.handleContainerStoppedTransitionError(event, container, currentKnownStatus)
+	case api.ContainerStatusNone:
+		fallthrough
+	case api.ContainerCreated:
+		// No need to explicitly stop containers if this is a * -> NONE/CREATED transition
+		seelog.Warnf("Managed task [%s]: Error creating container [%s]; marking its desired status as STOPPED: %v",
 			mtask.Arn, container.Name, event.Error)
-		// Leave the container known status as it is when encountered transition error,
-		// as we are not sure if the container status changed or not, we will get the actual
-		// status change from the docker event stream
 		container.SetKnownStatus(currentKnownStatus)
 		container.SetDesiredStatus(api.ContainerStopped)
+		return false
+	default:
+		// If this is a * -> RUNNING / RESOURCES_PROVISIONED transition, we need to stop
+		// the container.
+		seelog.Warnf("Managed task [%s]: Error starting/provisioning container[%s]; marking its desired status as STOPPED: %v",
+			mtask.Arn, container.Name, event.Error)
+		container.SetKnownStatus(currentKnownStatus)
+		container.SetDesiredStatus(api.ContainerStopped)
+		errorName := event.Error.ErrorName()
+		if errorName == dockerTimeoutErrorName || errorName == cannotInspectContainerErrorName {
+			// If there's an error with inspecting the container or in case of timeout error,
+			// we'll also assume that the container has transitioned to RUNNING and issue
+			// a stop
+			seelog.Warnf("Managed task [%s]: Forcing container [%s] to stop",
+				mtask.Arn, container.Name)
+			go mtask.engine.transitionContainer(mtask.Task, container, api.ContainerStopped)
```

Comment:
Do we need to check the `KnownStatus` of the container here? In case it has been stopped or already running, eg:
```
StartContainer -> inspect timeout -> stop the container 
StartEvent -> inspect timeout -> stop the container
```